# **Combinational Digital Circuit Solver** 

## **Sort nets in topogogical order**

In [1]:
import networkx as nx

def netlist(file):
    edges = []
    dictt = {}
    # Reading from the file
    with open(file,'r') as file1:
        each = file1.readlines()
        for l in each:
              
            # Creating a dictionary to keep track of inputs and gates connected to a node
            if(l.split()[1] != 'inv' and l.split()[1] != 'buf'):
                edges.append((l.split()[2],l.split()[4]))
                edges.append((l.split()[3],l.split()[4]))
                dictt[l.split()[4]] = [l.split()[2],l.split()[3],l.split()[1]]

            else:
                edges.append((l.split()[2],l.split()[3]))
                dictt[l.split()[3]] = [l.split()[2],l.split()[1]]

    net = nx.DiGraph()
    net.add_edges_from(edges)
    tsort = []
    alsort = []
    max_length = 0

    # Checking if the graph contains cycles
    if (nx.is_directed_acyclic_graph(net) == True):
        tsort = list(nx.topological_sort(net))
        max_length = len(max(tsort,key=len))
        alsort = tsort.copy()
        alsort.sort()

    return tsort,max_length,alsort,dictt,net



### Printing all given Nets in topological order

In [2]:
tsort,max_length,alsort,dictt,net = netlist("c8.net")
if(len(tsort)):
  print("The c8.net in topological order: \n")
  print(tsort)
else: print("The graph contains cycles")


The c8.net in topological order: 

['q', 'h', 'g', 'f', 'bn0', 'e', 'an0', 'd', 'r', 'cn0', 'y', 'z', 'c', 'b', 'a', 'v', 'w', 'x', 'i', 'l', 'k', 'm', 'n', 'j', 'p', 'u', 'o', 's', 'n_1', 'n_0', 'dummy_11', 'dummy_32', 'dummy_37', 'n_20', 'ln0', 'dummy_28', 'dummy_12', 'dummy_30', 'n_43', 'dummy_39', 'n_2', 'dummy_21', 'dummy_26', 'dummy_24', 'n_31', 'dummy_34', 'dummy_36', 'n_5', 'n_16', 'n_60', 'n_7', 'dummy_25', 'dummy_27', 'dummy_29', 'dummy_31', 'dummy_33', 'dummy_35', 'dummy_38', 'dummy_40', 'dummy_23', 'n_6', 'n_25', 'n_23', 'n_17', 'n_14', 'n_12', 'n_11', 'n_10', 'n_9', 'n_8', 'dummy_1', 'dummy_3', 'dummy_5', 'dummy_7', 'dummy_10', 'dummy_15', 'dummy_18', 'dummy_20', 'dummy_22', 'gn0', 'fn0', 'hn0', 'in0', 'kn0', 'dn0', 'jn0', 'en0', 'n_24', 'n_27', 'n_26', 'n_32', 'n_28', 'dummy_19', 'dummy_17', 'dummy_16', 'n_35', 'n_30', 'n_33', 'n_29', 'n_34', 'n_37', 'n_36', 'n_44', 'dummy_14', 'dummy_13', 'mn0', 'nn0', 'dummy_9', 'dummy_8', 'n_48', 'n_42', 'n_39', 'n_40', 'n_47', 'n_50',

In [3]:
tsort,max_length,alsort,dictt,net = netlist("c17_1.net")
if(len(tsort)):
  print("The c17_1.net in topological order: \n")
  print(tsort)
else: print("The graph contains cycles")

The graph contains cycles


In [4]:
tsort,max_length,alsort,dictt,net = netlist("c17.net")
if(len(tsort)):
  print("The c17.net in topological order: \n")
  print(tsort)
else: print("The graph contains cycles")

The c17.net in topological order: 

['N2', 'N7', 'N1', 'N3', 'N6', 'n_0', 'n_1', 'n_3', 'n_2', 'N22', 'N23']


In [5]:
tsort,max_length,alsort,dictt,net = netlist("c432.net")
if(len(tsort)):
  print("The c432.net in topological order: \n")
  print(tsort)
else: print("The graph contains cycles")

The c432.net in topological order: 

['N4', 'N108', 'N69', 'N17', 'N30', 'N56', 'N82', 'N95', 'N92', 'N76', 'N79', 'N63', 'N27', 'N11', 'N66', 'N50', 'N105', 'N89', 'N40', 'N24', 'N14', 'N53', 'N8', 'N1', 'N47', 'N86', 'N73', 'N99', 'N34', 'N21', 'N60', 'N43', 'N112', 'N115', 'N37', 'N102', 'n_2', 'n_4', 'n_8', 'n_3', 'n_1', 'n_9', 'n_5', 'n_0', 'n_6', 'n_7', 'n_20', 'n_12', 'n_10', 'n_13', 'n_14', 'dummy_37', 'n_15', 'n_21', 'n_19', 'dummy_56', 'n_11', 'dummy_58', 'dummy_57', 'dummy_54', 'n_17', 'n_16', 'dummy_55', 'n_18', 'n_22', 'n_23', 'n_24', 'n_37', 'n_39', 'n_34', 'n_27', 'n_31', 'n_43', 'n_26', 'n_29', 'N223', 'n_42', 'n_38', 'n_41', 'n_40', 'n_36', 'n_35', 'n_33', 'n_28', 'dummy_35', 'n_32', 'n_44', 'dummy_36', 'n_46', 'n_30', 'dummy_16', 'dummy_18', 'dummy_20', 'dummy_22', 'dummy_24', 'dummy_26', 'dummy_40', 'n_25', 'n_59', 'n_58', 'n_57', 'n_56', 'n_48', 'n_52', 'n_54', 'n_53', 'n_63', 'n_50', 'dummy_48', 'n_47', 'n_62', 'n_45', 'dummy_49', 'dummy_43', 'dummy_50', 'n_49', 'd

In [6]:
tsort,max_length,alsort,dictt,net = netlist("parity.net")
if(len(tsort)):
  print("The parity.net in topological order: \n")
  print(tsort)
else: print("The graph contains cycles")

The parity.net in topological order: 

['o', 'p', 'k', 'l', 'c', 'd', 'h', 'g', 'f', 'e', 'n', 'm', 'j', 'i', 'b', 'a', 'dummy_2', 'dummy_3', 'dummy_4', 'n_4', 'n_3', 'n_2', 'n_1', 'n_0', 'dummy_1', 'n_7', 'n_6', 'n_5', 'dummy_0', 'n_8', 'q']


<br>

## **Evaluation of circuit**

### 1) Using topological sort and multiple rounds of circuit evaluations

In [7]:
def topological_sort(filename):
   with open(filename,'r') as file2:
    get = file2.readlines()
    row1 = get[0].strip().split()
    pi = len(row1)
    writee = open(f"{filename}_output.txt",'w')
    #Printing the nodes in alphabetical order
    for i in range(0,len(tsort)): writee.write(f"{alsort[i]:<{max_length}} ")
    writee.write("\n")

    for i in range(1,len(get)):
        row = list(map(int,get[i].strip().split()))
        mapp = dict(zip(row1,row))

        # Going through the nodes from lower level to higher level
        for j in range(pi,len(tsort)):

          # Doing calculation as per the gate connected to find output
          if(len(dictt[tsort[j]])==3):
            if (dictt[tsort[j]][2]=="nand2"): mapp[tsort[j]] = (mapp[dictt[tsort[j]][1]]*mapp[dictt[tsort[j]][0]])^1
            if (dictt[tsort[j]][2]=="and2"): mapp[tsort[j]] = (mapp[dictt[tsort[j]][1]]*mapp[dictt[tsort[j]][0]])
            if (dictt[tsort[j]][2]=="or2"): mapp[tsort[j]] = (mapp[dictt[tsort[j]][1]]|mapp[dictt[tsort[j]][0]])
            if (dictt[tsort[j]][2]=="nor2"): mapp[tsort[j]] = (mapp[dictt[tsort[j]][1]]|mapp[dictt[tsort[j]][0]])^1
            if (dictt[tsort[j]][2]=="xor2"): mapp[tsort[j]] = (mapp[dictt[tsort[j]][1]]^mapp[dictt[tsort[j]][0]])
            if (dictt[tsort[j]][2]=="xnor2"): mapp[tsort[j]] = (mapp[dictt[tsort[j]][1]]^mapp[dictt[tsort[j]][0]])^1
          else:
            if (dictt[tsort[j]][1]=="inv"): mapp[tsort[j]] = (mapp[dictt[tsort[j]][0]])^1
            if (dictt[tsort[j]][1]=="buf"): mapp[tsort[j]] = (mapp[dictt[tsort[j]][0]])
          
        #Printing by formatting so that the output alingns with nodes
        for i in range(len(alsort)):
           writee.write(f"{mapp[alsort[i]]:<{max_length}} ")
        writee.write("\n")


### Few sample netlist and their ouput using topological sort

In [8]:
tsort,max_length,alsort,dictt,net = netlist("test.net")
topological_sort("test.inputs")

*Output for the sample netlist given in the presentation :*
```
A B C D E F G H I J 
0 1 1 0 0 0 0 0 0 0 
1 1 1 1 0 0 0 0 1 0
```

In [9]:
tsort,max_length,alsort,dictt,net = netlist("c17.net")
topological_sort("c17.inputs")

*Output for the c17 netlist:*
```
N1  N2  N22 N23 N3  N6  N7  n_0 n_1 n_2 n_3 
0   1   1   1   0   0   0   1   1   1   0   
0   0   0   0   1   0   0   1   1   1   1   
1   0   0   0   0   0   0   1   1   1   1   
0   0   0   0   1   1   1   1   0   1   1   
1   1   1   0   1   1   1   0   0   1   1   
1   1   1   1   1   0   0   0   1   1   0   
1   1   1   0   1   1   0   0   0   1   1   
1   1   1   1   0   0   0   1   1   1   0   
0   1   1   1   1   0   1   1   1   0   0   
0   0   0   0   1   1   0   1   0   1   1   
```

### 2) Using event driven simulation

In [33]:
from queue import Queue
#Initializing the queue
def event_driven(file):
  que = Queue()
  ini = [-1]*len(tsort)
  with open(file,'r') as file2:
    get = file2.readlines()
    row1 = get[0].strip().split()

    writee = open(f"{file}_output.txt",'w')
    for i in range(0,len(tsort)): writee.write(f"{alsort[i]:<{max_length}} ")
    writee.write("\n")
    
    state = dict(zip(tsort,ini))
    for i in range(1,len(get)):
        row = list(map(int,get[i].strip().split()))
        mapp = dict(zip(row1,row))
        track = [-1]*len(tsort) 
        repeat = dict(zip(tsort,track))  
        for pnode in row1:
            if (mapp[pnode] != state[pnode]):  
                # Adding nodes to queue whose state needs to be updated
                que.put(pnode)
                repeat[pnode] = 1
        
        while(not que.empty()):
            # Removing the front element
            remove = que.get()
            # Updating the state of nodes
            if(repeat[remove]==1):
                if remove in mapp: state[remove] = mapp[remove]
                else:
                    # Doing calculation as per the gate connected to find output
                    if(len(dictt[remove])==3):
                        if(repeat[dictt[remove][0]] == -1 and repeat[dictt[remove][1]] == -1):
                            if (dictt[remove][2]=="nand2"): state[remove] = (state[dictt[remove][1]]*state[dictt[remove][0]])^1
                            if (dictt[remove][2]=="and2"): state[remove] = (state[dictt[remove][1]]*state[dictt[remove][0]])
                            if (dictt[remove][2]=="or2"): state[remove] = (state[dictt[remove][1]]|state[dictt[remove][0]])
                            if (dictt[remove][2]=="nor2"): state[remove] = (state[dictt[remove][1]]|state[dictt[remove][0]])^1
                            if (dictt[remove][2]=="xor2"): state[remove] = (state[dictt[remove][1]]^state[dictt[remove][0]])
                            if (dictt[remove][2]=="xnor2"): state[remove] = (state[dictt[remove][1]]^state[dictt[remove][0]])^1
                    else: 
                            if(repeat[dictt[remove][0]] == -1):
                                if (dictt[remove][1]=="inv"): state[remove] = (state[dictt[remove][0]])^1
                                if (dictt[remove][1]=="buf"): state[remove] = (state[dictt[remove][0]])
                repeat[remove] = -1
                        
                list(map(que.put,list(net.successors(remove))))
                for node in list(net.successors(remove)): repeat[node] =1

        #Printing by formatting so that the output alingns with nodes     
        for i in range(len(alsort)):
          writee.write(f"{state[alsort[i]]:<{max_length}} ")
        writee.write("\n")


### Few sample netlist and their ouput using Event driven simulation

In [34]:
tsort,max_length,alsort,dictt,net = netlist("test.net")
event_driven("test.inputs")

*Output for the test netlist :*
```
A B C D E F G H I J 
0 1 1 0 0 0 0 0 0 0 
1 1 1 1 0 0 0 0 1 0 
```

In [35]:
tsort,max_length,alsort,dictt,net = netlist("c17.net")
event_driven("c17.inputs")

*Output for the c17 netlist :*
```
N1  N2  N22 N23 N3  N6  N7  n_0 n_1 n_2 n_3 
0   1   1   1   0   0   0   1   1   1   0   
0   0   0   0   1   0   0   1   1   1   1   
1   0   0   0   0   0   0   1   1   1   1   
0   0   0   0   1   1   1   1   0   1   1   
1   1   1   0   1   1   1   0   0   1   1   
1   1   1   1   1   0   0   0   1   1   0   
1   1   1   0   1   1   0   0   0   1   1   
1   1   1   1   0   0   0   1   1   1   0   
0   1   1   1   1   0   1   1   1   0   0   
0   0   0   0   1   1   0   1   0   1   1   
```


## **Which method is faster?**

#### *Test case - c8.net*

In [36]:
tsort,max_length,alsort,dictt,net = netlist("c8.net")
print("Number of nodes in the graph: ",end = "")
print(len(tsort))
if(len(tsort)):

    
    print(f"Time of execution for Event driven method: ")
    %timeit event_driven("c8.inputs")

    print("\n")

    print(f"Time of execution for topological sort method: ")
    %timeit topological_sort("c8.inputs")


else: print("The graph contains cycles")


Number of nodes in the graph: 135
Time of execution for Event driven method: 
5.56 ms ± 141 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


Time of execution for topological sort method: 
1.28 ms ± 66 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### *Test case - c17.net*

In [30]:
tsort,max_length,alsort,dictt,net = netlist("c17.net")
print("Number of nodes in the graph: ",end = "")
print(len(tsort))
if(len(tsort)):

    print(f"Time of execution for topological sort method: ")
    %timeit topological_sort("c17.inputs")

    print("\n")

    print(f"Time of execution for Event driven method: ")
    %timeit event_driven("c17.inputs")

else: print("The graph contains cycles")

Number of nodes in the graph: 11
Time of execution for topological sort method: 
641 µs ± 91.2 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


Time of execution for Event driven method: 
1.11 ms ± 133 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### *Test case - c432.net*

In [29]:
tsort,max_length,alsort,dictt,net = netlist("c432.net")
print("Number of nodes in the graph: ",end = "")
print(len(tsort))
if(len(tsort)):

    print(f"Time of execution for topological sort method: ")
    %timeit topological_sort("c432.inputs")

    print("\n")

    print(f"Time of execution for Event driven method: ")
    %timeit event_driven("c432.inputs")

else: print("The graph contains cycles")

Number of nodes in the graph: 219
Time of execution for topological sort method: 
24.8 ms ± 485 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


Time of execution for Event driven method: 
616 ms ± 11.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


#### *Test case - parity.net*

In [31]:
tsort,max_length,alsort,dictt,net = netlist("parity.net")
print("Number of nodes in the graph: ",end = "")
print(len(tsort))
if(len(tsort)):

    print(f"Time of execution for topological sort method: ")
    %timeit topological_sort("parity.inputs")

    print("\n")

    print(f"Time of execution for Event driven method: ")
    %timeit event_driven("parity.inputs")

else: print("The graph contains cycles")

Number of nodes in the graph: 31
Time of execution for topological sort method: 
706 µs ± 50.9 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


Time of execution for Event driven method: 
1.26 ms ± 44.2 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### *Test case - test.net*

In [37]:
tsort,max_length,alsort,dictt,net = netlist("test.net")
print("Number of nodes in the graph: ",end = "")
print(len(tsort))
if(len(tsort)):

    print(f"Time of execution for topological sort method: ")
    %timeit topological_sort("test.inputs")

    print("\n")

    print(f"Time of execution for Event driven method: ")
    %timeit event_driven("test.inputs")

else: print("The graph contains cycles")

Number of nodes in the graph: 10
Time of execution for topological sort method: 
549 µs ± 88.2 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


Time of execution for Event driven method: 
596 µs ± 25.4 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


<br>

## **Conclusion**

- When number of nodes of the graph is 219:
   - Topological sort method takes around 26 ms
   - Event driven simulation method takes around 642 ms
- When number of nodes of the graph is 10:
   - Topological sort method takes around 626 µs
   - Event driven simulation method takes around 689 µs
<br></br>
- From this it is quite obvious that, when the graph is small both the methods almost takes same time to execute but as the size of the graph increses the gap between their time of execution widens. That is Event driven takes longer time thus becoming less efficient for huge graphs compared to topological sort method. 

- From the state table it is very clear that, when the input does not change drastically, the state table won't alter much, and hence lesser time for computation. This facts also justifies the reason behind very low time of execution of test.net as just one primary input which is A has changed while others remains the same
